### ISSUE TITLE
MD020: Breaking &nbps; not detected inbetween heading hashtags and heading title

### ISSUE THREAD

**Issue by muttoni** [DavidAnson/markdownlint#367]
Hi, I had a devilishly difficult to track markdown issue caused by a weird space character inbetween the heading hashtags and the heading title. This was all due to the decimal charcode being 160 (non breaking space `&nbsp;`) instead of 32 (a regular space). 

```
---

## Subheading that doesn't work

## Subheading that works
---
```
---

## Subheading that works

---

Could you please detect this as an issue? As you can see, it breaks the Github markdown parser as well and it's very hard to track down unless you know the issue. I was blaming the surrounding contents for breaking the parser as Markdown Lint does not flag this as an issue (see below).

![image](https://user-images.githubusercontent.com/27052451/105038939-3e173280-5a60-11eb-9fd4-bf3230d43557.png)

**Edit**: I initially thought this could be a rare edge case, but I've now realized what caused the nbps...typing with an IT layout means that in order to type a hashtag I need to press the <kbd>Alt Gr</kbd> modifier key (+ <kbd>à</kbd>). The modifier key also causes <kbd>spacebar</kbd> to generate a non-breaking space. When typing quickly it's easy to press <kbd>spacebar</kbd> while <kbd>Alt Gr</kbd> is still pressed.

**Comment by muttoni** [comment]
Upon browsing the source code, this is likely falls in the **MD018** category and a check probably needs to go here (aside from relevant tests being created): 

https://github.com/DavidAnson/markdownlint/blob/82cf68023f7dbd2948a65c53fc30482432195de4/lib/md018.js#L13-L18

It also might be a good time to extend the test across all types of non-admissable whitespace characters:

```
U+0009 Horizontal tab (HT)
U+00A0 Non-break space
U+1680 Ogham space mark
U+180E Mongolian vowel separator
U+2000 En quad
U+2001 Em quad
U+2002 En space
U+2003 Em space
U+2004 Three-per-em space
U+2005 Four-per-em space
U+2006 Six-per-em space
U+2007 Figure space
U+2008 Punctuation space
U+2009 Thin space
U+200A Hair space
U+202F Narrow no-break space
U+205F Medium mathematical space
U+3000 Ideographic space
```

**Comment by DavidAnson** [comment]
This looks to be because the CommonMark definition of "whitespace" is more restrictive than JavaScript's RegExp:

- https://spec.commonmark.org/0.29/#characters-and-lines
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Character_Classes
